docs/rl-algorithms/rpo.md

```diff
@@ -0,0 +1,129 @@
+| dm_control/manipulator-bring_peg-v0   | 0.57 ± 0.24                                                  | 2.64 ± 2.19                                   |
+| dm_control/manipulator-insert_ball-v0 | 41.53 ± 15.58                                                | 39.24 ± 17.19                                 |
+| dm_control/manipulator-insert_peg-v0  | 49.89 ± 14.26                                                | 53.23 ± 16.02                                 |
+| dm_control/pendulum-swingup-v0        | 464.82 ± 379.09                                              | 776.21 ± 20.15                                |
+| dm_control/point_mass-easy-v0         | 530.17 ± 262.14                                              | 652.18 ± 21.93                                |
+| dm_control/point_mass-hard-v0         | 132.29 ± 69.71                                               | 184.29 ± 24.17                                |
+| dm_control/quadruped-walk-v0          | 239.11 ± 85.20                                               | 600.71 ± 213.40                               |
+| dm_control/quadruped-run-v0           | 165.70 ± 43.21                                               | 375.78 ± 119.95                               |
+| dm_control/quadruped-escape-v0        | 23.82 ± 10.82                                                | 68.19 ± 30.69                                 |
+| dm_control/quadruped-fetch-v0         | 183.03 ± 37.50                                               | 227.47 ± 21.67                                |
+| dm_control/reacher-easy-v0            | 769.10 ± 48.35                                               | 740.93 ± 51.58                                |
+| dm_control/reacher-hard-v0            | 636.89 ± 77.68                                               | 584.33 ± 43.76                                |
+| dm_control/stacker-stack_2-v0         | 61.73 ± 17.34                                                | 64.32 ± 7.15                                  |
+| dm_control/stacker-stack_4-v0         | 71.68 ± 29.03                                                | 53.78 ± 20.09                                 |
+| dm_control/swimmer-swimmer6-v0        | 152.82 ± 30.77                                               | 166.89 ± 14.49                                |
+| dm_control/swimmer-swimmer15-v0       | 148.59 ± 22.63                                               | 153.45 ± 17.12                                |
+| dm_control/walker-stand-v0            | 453.73 ± 217.72                                              | 736.72 ± 145.16                               |
+| dm_control/walker-walk-v0             | 308.91 ± 87.85                                               | 785.61 ± 133.29                               |
+| dm_control/walker-run-v0              | 123.98 ± 91.34                                               | 384.08 ± 123.89                               |
```

Comment:
A couple of comments:

1. The table formatting is broken at https://cleanrl-git-fork-masud99r-master-vwxyzjn.vercel.app/rl-algorithms/rpo/#experiment-results?
2. When displaying the gymnasium results, we usually display them in tabs such as https://docs.cleanrl.dev/rl-algorithms/ppo/#experiment-results_2. See the source code at https://github.com/vwxyzjn/cleanrl/blob/master/docs/rl-algorithms/ppo.md#experiment-results-2.
3. Do you have the mujoco results on `HalfCheetah-v4 Walker2d-v4 Hopper-v4 InvertedPendulum-v4 Humanoid-v4 Pusher-v4`?

Reply:
Thanks for the thorough feedback. I will make sure to update the docs accordingly (1 & 2).

3. As for Mujoco results, I still need to run them and see how RPO performs. So I will post the results here and update the docs. These results will help in deciding the depreciation decision of  `ppo_continuous_action.py`. 

Typically, RPO shows improved results compared to PPO in difficult environments. It seems PPO is performing quite well already on many Mujoco environments. I will start running the experiments to see how the performance stands.

Reply:
Where can I find a list of all the mujoco v2 and v4 environments? @vwxyzjn

Reply:
That makes sense. I think in the mujoco environments, RPO should not make the performance worse. Let's run the v2 environments as well just to be sure:

```
poetry install --with mujoco_py,mujoco
poetry run python -c "import mujoco_py"
OMP_NUM_THREADS=1 xvfb-run -a poetry run python -m cleanrl_utils.benchmark \
    --env-ids HalfCheetah-v2 Walker2d-v2 Hopper-v2 InvertedPendulum-v2 Humanoid-v2 Pusher-v2 \
    --command "poetry run python cleanrl/ppo_continuous_action.py --cuda False --track --capture-video" \
    --num-seeds 3 \
    --workers 1
poetry install --with mujoco
OMP_NUM_THREADS=1 xvfb-run -a python -m cleanrl_utils.benchmark \
    --env-ids HalfCheetah-v4 Walker2d-v4 Hopper-v4 InvertedPendulum-v4 Humanoid-v4 Pusher-v4 \
    --command "poetry run python cleanrl/gymnasium_support/ppo_continuous_action.py --cuda False --track" \
    --num-seeds 3 \
    --workers 1
```

The full mujoco environments are at https://gymnasium.farama.org/environments/mujoco/. Feel free to run all of them :), which would be


Ant-v2 HalfCheetah-v2 Hopper-v2 HumanoidStandup-v2 Humanoid-v2 InvertedDoublePendulum-v2 InvertedPendulum-v2 Reacher-v2 Swimmer-v2 Pusher-v2 Walker2d-v2  

and 

Ant-v4 HalfCheetah-v4 Hopper-v4 HumanoidStandup-v4 Humanoid-v4 InvertedDoublePendulum-v4 InvertedPendulum-v4 Reacher-v4 Swimmer-v4 Pusher-v4 Walker2d-v4

Reply:
Talking about environments, would you like to benchmark results on PyBullet? Do you have a preferred list of environments? Here is a list I find: https://github.com/bulletphysics/bullet3/blob/master/examples/pybullet/gym/pybullet_envs/__init__.py

Reply:
I wouldn't say no to the pybullet experiments, but previously they were omitted because pybullet seemed to be no longer maintained. It's also the case that it needs an adapter for gymnasium 

https://github.com/bulletphysics/bullet3/issues/4369, https://github.com/bulletphysics/bullet3/pull/4335. However, it is possible to fix this issue with a compatibility wrapper (https://gymnasium.farama.org/api/wrappers/misc_wrappers/#gymnasium.wrappers.EnvCompatibility).  

If you would like to add the pybullet results, feel free to use your preferred list of experiments and add another tab in the experiment section to report the results. 
<img width="783" alt="image" src="https://user-images.githubusercontent.com/5555347/209744917-6cf9d08c-238c-4edf-b91c-4c415d640b85.png">

Reply:
Results are in for mujoco (v4, and v2). Some environments require tuning of RPO `alpha` parameter as described below:

**Mujoco v4:**

Table:

|                     | ppo_continuous_action_8M ({'tag': ['v1.0.0-13-gcbd83f6']})   | rpo_continuous_action ({'tag': ['pr-331']})   |
|:--------------------|:-------------------------------------------------------------|:----------------------------------------------|
| HumanoidStandup-v4  | 107841.91 ± 15232.17                                         | 150129.37 ± 6447.36                           |
| Humanoid-v4         | 590.33 ± 29.38                                               | 763.79 ± 137.91                               |
| InvertedPendulum-v4 | 883.61 ± 25.01                                               | 875.56 ± 39.68                                |
| Walker2d-v4         | 2872.50 ± 683.82                                             | 3707.25 ± 346.22                              |

Learning curves:

![mujoco_v4_part1](https://user-images.githubusercontent.com/15100273/209876730-da7cde5f-1260-4009-96dd-8bbb563c40a8.png)

**Mujoco v2:**

Table:

|                     | ppo_continuous_action_8M ({'tag': ['v1.0.0-13-gcbd83f6']})   | rpo_continuous_action ({'tag': ['pr-331']})   |
|:--------------------|:-------------------------------------------------------------|:----------------------------------------------|
| HumanoidStandup-v2  | 108982.78 ± 17996.44                                         | 156755.15 ± 10672.52                          |
| Humanoid-v2         | 586.11 ± 44.21                                               | 723.01 ± 103.07                               |
| InvertedPendulum-v2 | 861.44 ± 27.30                                               | 857.88 ± 30.67                                |
| Walker2d-v2         | 3219.50 ± 948.84                                             | 3934.37 ± 428.82                              |

Learning curves:

![mujoco_v2_part1](https://user-images.githubusercontent.com/15100273/209876825-33bfcedf-5afd-4878-a117-5833d2ce965e.png)


The following environments require tuning of `alpha` (Algorithm 1, line 13, paper: https://arxiv.org/pdf/2212.07536.pdf). As described in the paper, this variable should be tuned for environments tested. A larger value means more randomness, whereas a smaller value indicates less randomness. Some mujoco environments require a smaller `alpha=0.01` value to achieve a reasonable performance compared to `alpha=0.5` for the rest of the environments. This version (`alpha=0.01`) of runs is indicated as `rpo_continuous_action_alpha_0_01` in the table and learning curves.

**Mujoco v4:**

Table:

|                           | ppo_continuous_action_8M ({'tag': ['v1.0.0-13-gcbd83f6']})   | rpo_continuous_action_alpha_0_01 ({'tag': ['pr-331']})   |
|:--------------------------|:-------------------------------------------------------------|:---------------------------------------------------------|
| Ant-v4                    | 1777.42 ± 844.67                                             | 2704.84 ± 665.91                                         |
| HalfCheetah-v4            | 2654.93 ± 1098.90                                            | 2729.61 ± 1311.41                                        |
| Hopper-v4                 | 2784.27 ± 280.23                                             | 2438.67 ± 492.42                                         |
| InvertedDoublePendulum-v4 | 5586.45 ± 213.17                                             | 5406.54 ± 269.84                                         |
| Reacher-v4                | -4.66 ± 0.84                                                 | -3.97 ± 0.17                                             |
| Swimmer-v4                | 124.45 ± 22.31                                               | 130.25 ± 12.48                                           |
| Pusher-v4                 | -30.54 ± 6.55                                                | -31.54 ± 9.80                                            |

Learning curves:

![mujoco_v4_part2](https://user-images.githubusercontent.com/15100273/209877045-c2d06cf7-c693-41aa-a763-3b4bc8592325.png)


**Mujoco v2:**

Table:

|                           | ppo_continuous_action_8M ({'tag': ['v1.0.0-13-gcbd83f6']})   | rpo_continuous_action_alpha_0_01 ({'tag': ['pr-331']})   |
|:--------------------------|:-------------------------------------------------------------|:---------------------------------------------------------|
| Ant-v2                    | 2435.03 ± 955.57                                             | 3142.98 ± 787.87                                         |
| HalfCheetah-v2            | 2714.14 ± 1260.34                                            | 2716.46 ± 1241.38                                        |
| Hopper-v2                 | 2411.13 ± 657.94                                             | 2286.05 ± 590.10                                         |
| InvertedDoublePendulum-v2 | 5727.86 ± 367.86                                             | 5811.94 ± 281.47                                         |
| Reacher-v2                | -4.57 ± 0.48                                                 | -4.17 ± 0.22                                             |
| Swimmer-v2                | 132.43 ± 10.01                                               | 141.49 ± 8.47                                            |
| Pusher-v2                 | -33.60 ± 8.54                                                | -26.08 ± 2.32                                            |

Learning curves:

![mujoco_v2_part2](https://user-images.githubusercontent.com/15100273/209877085-0f499aad-3117-4bf7-8ee2-2339f13c7c0b.png)

Reply:
As for the PyBullet, let's skip the experiments for now.

Reply:
These results look very nice! Did you run the utility with `--scan-history` for the v2 charts? The charts looked way too smooth compared to the v4 ones.

Reply:
I keep all the arguments the same. However, I had to uncomment ` n_samples=500,`, otherwise, I was getting an error message. Do you think that might cause the smoothness? 
Here is the error message:
```    uniques, codes = table.factorize(
  File "pandas/_libs/hashtable_class_helper.pxi", line 5943, in pandas._libs.hashtable.PyObjectHashTable.factorize
  File "pandas/_libs/hashtable_class_helper.pxi", line 5857, in pandas._libs.hashtable.PyObjectHashTable._unique
TypeError: unhashable type: 'dict'```

Reply:
I just ran the following and it gave me 
```
python -m openrlbenchmark.rlops \
    --filters '?we=openrlbenchmark&wpn=cleanrl&ceik=env_id&cen=exp_name&metric=charts/episodic_return' 'ppo_continuous_action_8M?tag=v1.0.0-13-gcbd83f6' 'rpo_continuous_action_alpha_0_01?tag=pr-331' \
    --env-ids HalfCheetah-v2 Walker2d-v2 Hopper-v2 InvertedPendulum-v2 Humanoid-v2 Pusher-v2 \
    --check-empty-runs False \
    --ncols 4 \
    --ncols-legend 3 \
    --output-filename compare \
    --scan-history
```

<img width="1467" alt="image" src="https://user-images.githubusercontent.com/5555347/210016522-ff94a6cf-8f07-4da3-be55-1d90b6e2c901.png">

Maybe try upgrading by doing `pip install openrlbenchmark --upgrade`